Step3 구현

guswns1659 opened the pull request:
Step3 구현 완료했습니다. 반복을 통해 스프링에 점점 익숙해지는 느낌입니다. 구현하며 배운점 기록했습니다. 

[배포URL](https://hyunjun2.herokuapp.com/)

## Step3 진행하며 배운 것 
- 자바지기 영상보니 모르면 검색하는 게 일상. 두려워하지 않고 검색한다. 
- Bad Request일 경우 mapping URL에 @RequestMapping에 해당하는 url이 있나 확인 ex)/users/{{id}}
- 사용자일 땐 몰랐던 기능들이 로그인 하나로 복잡해질 수 있음을 깨달음. 
- getter, setter로 객체의 속성에 직접 접근하지 말고 메세지를 보내서 자율성을 높이는 중요성을 깨달음. 과거 호눅스가 getter가 적으면 좋다는 피드백을 했는데 이해가 된다. 
- @ManyToOne은 Many에 해당하는 객체에 매핑을 한다.

## 해결 못한 것 
- auto reload 기능 : 코드 변경하면 서버 재시작하지 않고 바로 반영하게 하고 싶은데 검색대로 했는데 기능 동작안함.

guswns1659 committed:
20.02.26) step2 관련 추가 학습, README에 추가

한 개 데이터 넘길 때 handlebar 사용법.
개인정보 수정 시 user의 update 메서드로 넘기기.

guswns1659 committed:
20.02.27) 로그인 기능 구현 완료

로그인 성공 시 sesstion에 로그인 아이디, 비밀번호 저장.

guswns1659 committed:
20.02.27) 로그인 상태에 따른 메뉴 처리 및 로그아웃 기능 구현

import.sql 사용해서 데이터 1개 초기화.
handlebars 속성 변경해 sesstion 값이 넘어갈 수 있도록 변경
로그아웃 시 session에 저장된 user 제거.20.02.27) 로그인 상태에 따른 메뉴 처리 및 로그아웃 기능 구현

guswns1659 committed:
20.02.27) 타인 개인정보 수정 금지 기능 구현

guswns1659 committed:
20.02.28) 설정 변경

guswns1659 committed:
20.02.28) Step2 리팩토링

Question의 date 타입을 String에서 LocalDateTime으로 변경

guswns1659 committed:
20.03.02) Step3 추가 리팩토링

- HttpSessionUtils 생성해 로그인 관련 메서드 생성.
- User가 직접 비밀번호, id를 확인하는 메서드 추가.
- Optional.orElseThrow 통해 예외 처리.

guswns1659 committed:
20.03.02) 질문하기 기능 구현

- HomeController 생성 후 index Controller 이동.
- 질문의 writer는 user의 이름으로 구현.

guswns1659 committed:
20.03.02) 질문 수정하기 기능 구현

- PutMapping 가능한 설정 추가.

guswns1659 committed:
20.03.02) 질문 삭제하기 기능 구현

guswns1659 committed:
20.03.02) User와 Question 연결 실습

- @ManyToOne 이용한 객체간 관계 정의.

guswns1659 committed:
20.03.03) 답변 기능 구현

- Answer 객체, AnswerRepository 생성.

guswns1659 committed:
20.03.03) 답변 수정, 삭제 기능 구현

- 질문 수정 및 삭제와 비슷한 로직으로 구현.

guswns1659 committed:
20.03.03) README에 구현 중 배운점, 어려운점 기록

guswns1659 committed:
20.03.03) Step4 에 있던 step3 진행과정 및 어려운 점 추가

guswns1659 committed:
20.03.04) 불필요한 공백 제거

LarryJung commented on src/main/java/com/codessquad/qna/AnswerController.java:
```diff
@@ -0,0 +1,100 @@
+    @Autowired
+    private QuestionRepository questionRepository;
+
+    @PostMapping("/{question.id}/answers")
```
보통 question.id 보다는 questionId 로 사용합니다.ㅎㅎ

guswns1659 replied:
넵! 참고하겠습니다.

LarryJung commented on src/main/java/com/codessquad/qna/AnswerController.java:
```diff
@@ -0,0 +1,100 @@
+
+    @PostMapping("/{question.id}/answers")
+    public String answer(@PathVariable("question.id") Long questionId,
+                         @RequestParam String contents,
```
컨텐츠가 String 하나 뿐이여서 RequestParam으로 사용하신 것처럼 생각됩니다. ㅎㅎ
내용물이 많으면(객체형태) requestBody로 전달하는 것 아시죠?

guswns1659 replied:
질문있습니다! 
내용물이 많으면(객체형태) requestBody로 전달하는 것의 의미가 현재 Question, User 객체처럼 넘어오는 데이터를 한번에 받는 방법이 맞을까요?

honux77 replied:
맞아요~

LarryJung commented on src/main/java/com/codessquad/qna/AnswerController.java:
```diff
@@ -0,0 +1,100 @@
+                             Model model) {
+        User sessionedUser = HttpSessionUtils.getUserFromSession(httpSession);
+
+        if (!HttpSessionUtils.isLoginUser(sessionedUser)) {
+            return "redirect:/users/loginForm";
+        }
```
이 부분 중복이 일어나는데 조금 더 줄여볼 수는 없을까요? 다양한 방법이 있을 건데 한번 고민해 보시는 걸 권해드려요

guswns1659 replied:
고민해보겠습니다.

LarryJung commented on src/main/java/com/codessquad/qna/AnswerController.java:
```diff
@@ -0,0 +1,100 @@
+        return "redirect:/questions/{question.id}";
+    }
+
+    private Answer findAnswer(AnswerRepository answerRepository, Long answerId) {
```
메서드 이름 findById로 repository 메서드 이름과 동일하게 사용하셔도 됩니다.^^

guswns1659 replied:
넵!

LarryJung commented on src/main/java/com/codessquad/qna/UserController.java:
```diff
@@ -54,17 +87,35 @@ public String updateUser(@PathVariable Long id,
+            return "redirect:/users/{id}/form";
+        }
+
+        if (newPassword.equals(checkPassword)) {
```
이전 패스워드를 requestParam으로 보내면 Url에 노출이 있으니 위험하다고 생각합니다. 대신에 findUser로 조회한 유저객체의 메서드를 활용하여 비밀번호를 체크하는 것은 어떨까요?

guswns1659 replied:
질문있습니다! 
1. Post방식으로 보내면 데이터가 엔티디에 담겨오니까 URI에 안보이지 않을까 해서 사용했습니다. Post로 사용해도 보안 문제가 있을까요??

LarryJung commented on src/main/java/com/codessquad/qna/domain/Answer.java:
```diff
@@ -0,0 +1,73 @@
+    private Long id;
+    private String writer;
+    private String contents;
+    private LocalDateTime date;
```
date는 딱 2019-03-03 처럼 일단위까지의 정보를 표현할 때 사용합니다! 분, 초 그 이하의 단위의 시간정보를 표현할때는 
createdTime, createdAt 등을 주로 사용합니다~ 더 좋은 표현이 없을지 고민해보시고 적용해주세요^^

guswns1659 replied:
네 CreatedTime이 적절한 것 같습니다!

LarryJung commented on src/main/java/com/codessquad/qna/domain/Answer.java:
```diff
@@ -0,0 +1,73 @@
+
+    public void update(String contents) {
+        this.contents = contents;
+        this.date = LocalDateTime.now();
```
update에 date를 새로 갱싱하는 것을 보면 date필드는 updatedTime, updatedAt 등이 적절할 것 같네요.
보통 식별자가 있는 테이블은 생성시각, 수정시각 컬럼은 같이 들어가는 편이니 참고해주세요!

guswns1659 replied:
질문있습니다! 
식별자가 있는 테이블의 의미가 무엇인가요??

LarryJung commented on src/main/java/com/codessquad/qna/domain/Answer.java:
```diff
@@ -0,0 +1,73 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private String writer;
```
Question 객체는 User객체가 연관관계로 들어가 있는데 Answer는 유저와 연관관계를 맺지 않은 이유가 있으신가요?

guswns1659 replied:
단순한 생각이지만 Join된 테이블이 많으면 결합도가 높아진다 생각했던 것 같습니다. 
그런데 피드백 주신 것을 생각해보니 테이블 간의 join이 객체의 결합도랑은 전혀 다른 개념같습니다. 
테이블 개수도 많지도 않고 연관관계를 맺는게 나을 것 같습니다.

LarryJung commented on src/main/java/com/codessquad/qna/domain/Question.java:
```diff
@@ -61,10 +71,16 @@ public void setContents(String contents) {
-                ", writer='" + writer + '\'' +
                 ", title='" + title + '\'' +
+                ", date=" + date +
+                ", writer='" + writer + '\'' +
```
연관관계를 맺기 시작하고부터는 toString메서드를 작성할 때 주의할 점이 있습니다. (한번 검색해보시는 걸 권해드립니다.ㅎㅎㅎ)

guswns1659 replied:
넵!

LarryJung commented on src/main/java/com/codessquad/qna/QuestionController.java:
```diff
@@ -1,43 +1,117 @@
-    @GetMapping("/questions/{id}")
-    public String questionPage(@PathVariable Long id, Model model) {
-        model.addAttribute("question", questionRepository.findById(id).get());
+        if (answers == null) {
```
repository에서 리스트 데이터가 없을 경우 기본은 emptyList가 반환 됩니다. 그래서 answers == null 이 아닌 answers.isEmpty() 로 체크해야 합니다. https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.nullability 
`repository methods returning collections, collection alternatives, wrappers, and streams are guaranteed never to return null but rather the corresponding empty representation.`

guswns1659 replied:
네! 올려주신 링크 들어가서 읽어보니 Collection<T> Query의 리턴 타입은 Collection<T>가 된다고 하네요! 그래서 null이 아니라 emptyList가 나오는지 이해됐습니다

LarryJung reviewed (commented):
수고 많으셨습니다! 꼼꼼하게 잘 구현하셨네요👍👍
몇가지 생각해볼 점들 리뷰 드렸습니다. 보시고 질문 있으시면 얼마든 환영입니다!

LarryJung commented:
넵 맞습니다

2020년 3월 7일 (토) 오후 12:53, DevJun <notifications@github.com>님이 작성:

> *@guswns1659* commented on this pull request.
> ------------------------------
>
> In src/main/java/com/codessquad/qna/AnswerController.java
> <https://github.com/code-squad/java-qna/pull/183#discussion_r389225140>:
>
> > +import org.springframework.ui.Model;
>
> +import org.springframework.web.bind.annotation.*;
>
> +
>
> +import javax.servlet.http.HttpSession;
>
> +
>
> +@Controller
>
> +@RequestMapping("/questions")
>
> +public class AnswerController {
>
> +    @Autowired
>
> +    private AnswerRepository answerRepository;
>
> +    @Autowired
>
> +    private QuestionRepository questionRepository;
>
> +
>
> +    @PostMapping("/{question.id}/answers")
>
> +    public String answer(@PathVariable("question.id") Long questionId,
>
> +                         @RequestParam String contents,
>
>
> 질문있습니다!
> 내용물이 많으면(객체형태) requestBody로 전달하는 것의 의미가 현재 Question, User 객체처럼 넘어오는 데이터를
> 한번에 받는 방법이 맞을까요?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/code-squad/java-qna/pull/183?email_source=notifications&email_token=AITJA55SOVCY2M3B4JJIYI3RGHANHA5CNFSM4K7XNUKKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCYML3NY#discussion_r389225140>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AITJA55MVSBLT5LWRUKG3VDRGHANHANCNFSM4K7XNUKA>
> .
>

LarryJung commented:
primary 컬럼이 있는 테이블을 말하고여, 식별자가 있다는 뜻은 그 데이터를 계속 추적(?) 하겠다는 의미이기도 하거든요,
조회를 하거나 조회를 해서 수정을 하거나..
그런 경우에는 생성시각, 수정시각 등이 여러가지 이유에서 많이 필요로 합니다.

2020년 3월 7일 (토) 오후 12:56, DevJun <notifications@github.com>님이 작성:

> *@guswns1659* commented on this pull request.
> ------------------------------
>
> In src/main/java/com/codessquad/qna/domain/Answer.java
> <https://github.com/code-squad/java-qna/pull/183#discussion_r389225312>:
>
> > +
>
> +    public void setDate(LocalDateTime date) {
>
> +        this.date = date;
>
> +    }
>
> +
>
> +    public Question getQuestion() {
>
> +        return question;
>
> +    }
>
> +
>
> +    public void setQuestion(Question question) {
>
> +        this.question = question;
>
> +    }
>
> +
>
> +    public void update(String contents) {
>
> +        this.contents = contents;
>
> +        this.date = LocalDateTime.now();
>
>
> 질문있습니다!
> 식별자가 있는 테이블의 의미가 무엇인가요??
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/code-squad/java-qna/pull/183?email_source=notifications&email_token=AITJA56LXPH3DKJEFKIMA53RGHAXTA5CNFSM4K7XNUKKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCYML4OQ#discussion_r389225312>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AITJA532IGVCXBPLKRJZWILRGHAXTANCNFSM4K7XNUKA>
> .
>

guswns1659 committed:
20.03.06) 리뷰어 피드백 기반 리팩토링

- Controller의 Mapping url 중 {question.id} -> {questionId} 변경
- date -> CreatedDate으로 변경
- Answer와 User 연결
- 답변 수정 시 전체 답변 보여주지 말고 해당 답변만 보이게 변경.
